use sync id when recovering

brwe opened the pull request:
Skip phase 1 of recovery in case an identical sync id was found on primary and replica. Relates to #10032

bleskes commented on src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -153,7 +155,13 @@ public InternalEngine(EngineConfig engineConfig, boolean skipInitialTranslogReco
                     recoverFromTranslog(translogId.v1(), transformer);
                 } else {
-                    flush(true, true);
+                    if (lastCommittedSegmentInfos.getUserData().get(SYNC_COMMIT_ID) == null) {
```
can we add a nocommit here plus a comment as to why we do this? I think we need to think about this some more but don't want to stop development now..

bleskes commented on src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -1263,9 +1274,15 @@ protected void recoverFromTranslog(long translogId, TranslogRecoveryPerformer ha
         if (operationsRecovered > 0) {
+            flush(true, true);
             refresh("translog recovery");
+        } else if (lastCommittedSegmentInfos.getUserData().get(SYNC_COMMIT_ID) == null) {
```
same comment about no commit.

bleskes commented on src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardRepository.java:
```diff
@@ -186,7 +186,8 @@ public IndexShardSnapshotStatus snapshotStatus(SnapshotId snapshotId, ShardId sh
     public void verify(String seed) {
-        BlobContainer testBlobContainer = blobStore.blobContainer(basePath);;
+        BlobContainer testBlobContainer = blobStore.blobContainer(basePath);
+        ;
```
extra semi-colon (was there before)

bleskes commented on src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -665,24 +666,30 @@ public String toString() {
             ImmutableMap.Builder<String, StoreFileMetaData> builder = ImmutableMap.builder();
             Map<String, String> checksumMap = readLegacyChecksums(directory).v1();
             try {
-                final SegmentInfos segmentCommitInfos = Store.readSegmentsInfo(commit, directory);
```
we can't take this one out of the try catch, it protects agains corruption. I think we should rename `buildMetadata` to `loadMetadata` and set both commutUserData and metadata in it.

bleskes commented on src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -969,9 +977,15 @@ public void readFrom(StreamInput in) throws IOException {
             for (int i = 0; i < size; i++) {
                 StoreFileMetaData meta = StoreFileMetaData.readStoreFileMetaData(in);
                 builder.put(meta.name(), meta);
+                logger.info("read {} {}", meta.name(), meta);
```
left over debugging tool?

bleskes commented on src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -969,9 +977,15 @@ public void readFrom(StreamInput in) throws IOException {
+            commitUserData = new HashMap<>();
+            int num = in.readVInt();
+            for (int i = num; i > 0; i--) {
+                commitUserData.put(in.readString(), in.readOptionalString());
```
we don't need the optional string here as we found out :)

bleskes commented on src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -980,6 +994,15 @@ public void writeTo(StreamOutput out) throws IOException {
+            out.writeVInt(commitUserData.size());
+            for (Map.Entry<String, String> entry : commitUserData.entrySet()) {
+                out.writeString(entry.getKey());
+                out.writeOptionalString(entry.getValue());
```
same here - no optional

bleskes commented on src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:
```diff
@@ -811,15 +812,7 @@ public synchronized long totalRecoverBytes() {
             return total;
         }
 
-        public synchronized long totalReuseBytes() {
```
the fact that this is not reused is actually a bug in the toXContent:

```
builder.byteSizeField(Fields.TOTAL_IN_BYTES, Fields.TOTAL, totalBytes());
 builder.byteSizeField(Fields.REUSED_IN_BYTES, Fields.REUSED, totalBytes());
            
```

can we keep this here? I'll make a pull request..

bleskes commented on src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -168,21 +170,33 @@ public void phase1(final SnapshotIndexCommit snapshot) throws ElasticsearchExcep
-            final Store.RecoveryDiff diff = recoverySourceMetadata.recoveryDiff(new Store.MetadataSnapshot(request.existingFiles()));
+            final Store.RecoveryDiff diff;
+            boolean copyFiles = true;
+            if (recoverySourceMetadata.getCommitUserData().get(Engine.SYNC_COMMIT_ID) != null &&
```
don't reuse a copyFile flag here as discussed.

bleskes commented on src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java:
```diff
@@ -46,7 +47,7 @@
     private boolean markAsRelocated;
 
-    private Map<String, StoreFileMetaData> existingFiles;
+    Store.MetadataSnapshot metadataSnapshot;
```
we lost the private?

bleskes commented on src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java:
```diff
@@ -126,10 +118,11 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeByte(recoveryType.id());
     }
+
+    public Store.MetadataSnapshot metadataSnapshot() {
```
can we put this next to the other getters?

bleskes commented on src/test/java/org/elasticsearch/gateway/RecoveryFromGatewayTests.java:
```diff
@@ -435,6 +439,65 @@ public void testReusePeerRecovery() throws Exception {
     @Test
     @Slow
+    @TestLogging("indices.recovery:TRACE,index.store:TRACE")
+    public void testSyncFlushedRecovery() throws Exception {
```
can we merge this test with the previous one where we randomly replace the first full cluster restart with an synced flush...

bleskes commented on src/test/java/org/elasticsearch/index/store/StoreTest.java:
```diff
@@ -713,9 +713,9 @@ public void testRecoveryDiffWithLegacyCommit() {
         metaDataMap.put("segments_1", new StoreFileMetaData("segments_1", 50, null, null, new BytesRef(new byte[] {1})));
         metaDataMap.put("_0_1.del", new StoreFileMetaData("_0_1.del", 42, "foobarbaz", null, new BytesRef()));
-        Store.MetadataSnapshot first = new Store.MetadataSnapshot(metaDataMap);
+        Store.MetadataSnapshot first = new Store.MetadataSnapshot(metaDataMap, Collections.EMPTY_MAP);
```
can we add a test somewhere here (or extend an existing one) to see we read commit data correctly?

brwe replied:
I added StoreTest.testUserDataRead(). Is this what you meant?

bleskes commented on src/test/java/org/elasticsearch/indices/recovery/StartRecoveryRequestTest.java:
```diff
@@ -69,11 +73,42 @@ public void testSerialization() throws Exception {
 
+    // TODO: where should this test be?
+    @Test
+    public void testMetadataSnapshotStreaming() throws Exception {
```
in StoreTest.java?

bleskes commented:
looks good. left some comments as discussed.

brwe commented:
Thanks for the review! I addressed all comments now. I found that GatewayAllocator has some heuristic built in that checks how much of a shard on a node coincides with the primary before assigning replicas and added a check for sync ids there as well. Let me know if that makes sense.

s1monw commented on src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -972,6 +976,11 @@ public void readFrom(StreamInput in) throws IOException {
             }
             this.metadata = builder.build();
             assert metadata.isEmpty() || numSegmentFiles() == 1 : "numSegmentFiles: " + numSegmentFiles();
+            commitUserData = new HashMap<>();
```
can this be immutable?

s1monw commented on src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -665,24 +664,29 @@ public String toString() {
 
         MetadataSnapshot(IndexCommit commit, Directory directory, ESLogger logger) throws IOException {
-            metadata = buildMetadata(commit, directory, logger);
+            loadMetadata(commit, directory, logger);
```
any chance we can return a tuple and make the two maps immutable?

brwe replied:
ok. also, made both metadata and commitUserData final and ImmutableMap

s1monw commented on src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -168,28 +169,43 @@ public void phase1(final SnapshotIndexCommit snapshot) throws ElasticsearchExcep
-                if (logger.isTraceEnabled()) {
-                    logger.trace("[{}][{}] recovery [phase1] to {}: not recovering [{}], exists in local store and has checksum [{}], size [{}]",
-                            indexName, shardId, request.targetNode(), md.name(), md.checksum(), md.length());
+            final boolean recoverWithSyncId = recoverySourceMetadata.getCommitUserData().get(Engine.SYNC_COMMIT_ID) != null &&
```
can you assign `recoverySourceMetadata.getCommitUserData().get(Engine.SYNC_COMMIT_ID)` to a local var makes it much easier to read

s1monw replied:
`request.metadataSnapshot().getCommitUserData()` this one as well please

s1monw commented on src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java:
```diff
@@ -155,12 +156,12 @@ private void doRecovery(final RecoveryStatus recoveryStatus) {
         assert recoveryStatus.sourceNode() != null : "can't do a recovery without a source node";
 
         logger.trace("collecting local files for {}", recoveryStatus);
-        Map<String, StoreFileMetaData> existingFiles;
```
good :)

s1monw commented on src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java:
```diff
@@ -109,13 +109,9 @@ public void readFrom(StreamInput in) throws IOException {
-            StoreFileMetaData md = StoreFileMetaData.readStoreFileMetaData(in);
-            existingFiles.put(md.name(), md);
-        }
+        metadataSnapshot = Store.MetadataSnapshot.read(in);
```
good!!

s1monw commented on src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java:
```diff
@@ -216,15 +218,15 @@ public ShardId shardId() {
         @Override
         public Iterator<StoreFileMetaData> iterator() {
-            return files.values().iterator();
+            return metadataSnapshot.asMap().values().iterator();
```
you can just return `metadataSnapshot` it impls iterable

s1monw commented on src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java:
```diff
@@ -237,22 +239,18 @@ public static StoreFilesMetaData readStoreFilesMetaData(StreamInput in) throws I
+        }
+
+        public String syncId() {
+            return metadataSnapshot.getCommitUserData().get(Engine.SYNC_COMMIT_ID);
```
maybe we add a utiltiy method to metadata snapshot ?

s1monw commented:
this looks very good I left some minor comments

s1monw commented:
> I found that GatewayAllocator has some heuristic built in that checks how much of a shard on a node coincides with the primary before assigning replicas and added a check for sync ids there as well. Let me know if that makes sense.

it does make a lot of sense to me

brwe commented:
Thanks for the review! Addressed all comments.

s1monw commented on src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -657,32 +657,42 @@ public String toString() {
      * @see StoreFileMetaData
      */
-    public final static class MetadataSnapshot implements Iterable<StoreFileMetaData>, Streamable {
+    public final static class MetadataSnapshot implements Iterable<StoreFileMetaData> {
```
we have a new interface `Writeable.java` that we can use here. It has:

```Java
    void writeTo(StreamOutput out) throws IOException;
    T readFrom(StreamInput in) throws IOException;
```
and it's usually implemented by passing the `StreamInput` to a dedicated ctor so everything can be final. Can we use that here too?

s1monw commented:
left one comment other than that LGTM

clintongormley commented:
@brwe can you add a description to the PR plus a link to the main issue?

brwe commented:
Pushed another commit.

s1monw commented on src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -1010,6 +1032,15 @@ private final int numSegmentFiles() { // only for asserts
             }
             return count;
         }
+
```
java docs?

s1monw commented on src/main/java/org/elasticsearch/indices/recovery/RecoveryCleanFilesRequest.java:
```diff
@@ -61,7 +61,7 @@ public void readFrom(StreamInput in) throws IOException {
         recoveryId = in.readLong();
         shardId = ShardId.readShardId(in);
-        snapshotFiles = Store.MetadataSnapshot.read(in);
+        snapshotFiles = new Store.MetadataSnapshot(in);
```
good

s1monw commented:
it looks good to me to go on the feature branch... I left two minor comments.

brwe commented:
added another commit with docs

s1monw commented:
++ go push

brwe commented:
pushed to https://github.com/elastic/elasticsearch/tree/feature/synced_flush
